### PR TITLE
[Clang][CMake] Support perf, LBR, and Instrument CLANG_BOLT options

### PR DESCRIPTION
**Pull request by aaupov**
Split up and refactor CLANG_BOLT_INSTRUMENT into support for
perf no-LBR and perf with LBR profiling modes.

Differential Revision: https://reviews.llvm.org/D143617

### CODE REVIEW

File: clang/utils/perf-training/bolt.lit.cfg [hunk 1]
```diff
@@ -6,15 +6,52 @@ import lit.util
 import os
 import subprocess
 
-config.clang = os.path.realpath(lit.util.which('clang-bolt.inst', config.clang_tools_dir)).replace('\\', '/')
+clang_binary = "clang"
+perf_wrapper = ""
+if config.clang_bolt_mode.lower() == "instrument":
+    clang_binary = config.clang_bolt_name
+else:  # perf or LBR
+    perf_wrapper = "%s %s/perf-helper.py perf" % (
```
Comment (petrhosek) [inline clang/utils/perf-training/bolt.lit.cfg:14]: Please use Python f-string.

File: clang/utils/perf-training/perf-helper.py [hunk 2]
```diff
@@ -67,6 +67,67 @@ def merge_fdata(args):
...
+        prog="perf-helper perf", description="perf wrapper for BOLT profile collection"
+    )
+    parser.add_argument(
+        "--lbr", action="store_true", help="Use perf with branch stacks"
+    )
+    parser.add_argument("cmd", nargs="*", help="")
+
+    # Use python's arg parser to handle all leading option arguments, but pass
+    # everything else through to perf
+    first_cmd = next(arg for arg in args if not arg.startswith("--"))
```
Comment (petrhosek) [inline clang/utils/perf-training/perf-helper.py:81]: You can use `parser.add_argument('cmd', nargs=argparse.REMAINDER)` to let `argparse` handle this for you.

File: clang/CMakeLists.txt [hunk 3]
```diff
@@ -850,23 +850,38 @@ if (CLANG_ENABLE_BOOTSTRAP)
   endforeach()
 endif()
 
-if (CLANG_BOLT_INSTRUMENT AND NOT LLVM_BUILD_INSTRUMENTED)
+set(CLANG_BOLT "INSTRUMENT" CACHE STRING "Apply BOLT optimization to Clang. \
+  May be specified as Instrument or Perf or LBR to use a particular profiling \
+  mechanism.")
+string(TOUPPER "${CLANG_BOLT}" uppercase_CLANG_BOLT)
```
Comment (petrhosek) [inline clang/CMakeLists.txt:856]: This is a super minor nit, but you can reuse the same variable (it wouldn't actually change the cached value, the non-cached variable is going to shadow the cached one which is fine in this case).

**Review by petrhosek — APPROVED** [review]

**Comment by dyung** [comment]
@aaupov I believe this change is somehow causing cmake failures on several bots I oversee, although I cannot explain why it passed when building your commit. For example:

https://lab.llvm.org/buildbot/#/builders/139/builds/57628
```
CMake Error at /home/buildbot/buildbot-root/llvm-clang-x86_64-sie-ubuntu-fast/llvm-project/clang/utils/perf-training/CMakeLists.txt:105 (add_custom_target):
  Error evaluating generator expression:
    $<TARGET_FILE:merge-fdata>
  No target "merge-fdata"
```
Any idea what might be going wrong here?

**Comment by DavidSpickett** [comment]
I think because bots don't always clean their cmake files, revert incoming.

**Comment by DavidSpickett** [comment]
https://lab.llvm.org/buildbot/#/builders/245/builds/19468

```
CMake Error at /home/tcwg-buildbot/worker/clang-armv8-quick/llvm/clang/utils/perf-training/CMakeLists.txt:105 (add_custom_target):
  Error evaluating generator expression:
    $<TARGET_FILE:merge-fdata>
  No target "merge-fdata"
```

This particular bot is clang+llvm and only the ARM target, but we've seen it on our AArch64 bots too.

File: clang/CMakeLists.txt [hunk 4]
```diff
@@ -850,23 +850,38 @@ if (CLANG_ENABLE_BOOTSTRAP)
   endforeach()
 endif()
 
-if (CLANG_BOLT_INSTRUMENT AND NOT LLVM_BUILD_INSTRUMENTED)
+set(CLANG_BOLT "INSTRUMENT" CACHE STRING "Apply BOLT optimization to Clang. \
```
Comment (DavidSpickett) [inline clang/CMakeLists.txt:853]: Wild guess, this defaults to `INSTRUMENT`, but it should default to some `OFF` value?

**Comment by nickdesaulniers** [comment]
I think the libc buildbots are now failing as a result of this change. @aaupov can you PTAL?

- https://lab.llvm.org/buildbot/#/builders/223/builds/34823
- https://lab.llvm.org/buildbot/#/builders/225/builds/29950

**Comment by aaupov** [comment]
Reverted. Strangely I couldn't repro from https://lab.llvm.org/buildbot/#/builders/225/builds/29950/ the failure locally. But will get to the bottom of it.

**Comment by aaupov** [comment]
These are all annotated builders. I tried triggering rebuilds with clean build dir through buildbot UI but it didn't have an effect. I don't think it's a reasonable expectation that every change must configure cleanly from an existing build directory.

**Comment by lntue** [comment]
> These are all annotated builders. I tried triggering rebuilds with clean build dir through buildbot UI but it didn't have an effect. I don't think it's a reasonable expectation that every change must configure cleanly from an existing build directory.

I'm sorry but I also don't think it is reasonable to re-landing a change that is known for breaking builders during the weekends, when not many people are around to react to the breakage.

**Comment by aaupov** [comment]
Status of the builders that reported breakage:
- [reverse-iteration](https://lab.llvm.org/buildbot/#/builders/54): https://lab.llvm.org/buildbot/#/builders/54/builds/8671 broken, https://lab.llvm.org/buildbot/#/builders/54/builds/8672 green, breakage was unrelated (unresolved tests)
- [sanitizer-aarch64-linux-bootstrap-msan](https://lab.llvm.org/buildbot/#/builders/237) - annotated builder: https://lab.llvm.org/buildbot/#/builders/237/builds/6271 broken, https://lab.llvm.org/buildbot/#/builders/237/builds/6272 green after cleanup step
- [sanitizer-x86_64-linux-bootstrap-ubsan](https://lab.llvm.org/buildbot/#/builders/85) - annotated builder: https://lab.llvm.org/buildbot/#/builders/85/builds/21896 broken, https://lab.llvm.org/buildbot/#/builders/85/builds/21897 green after cleanup step
- [sanitizer-aarch64-linux-bootstrap-ubsan](https://lab.llvm.org/buildbot/#/builders/238) - annotated builder: https://lab.llvm.org/buildbot/#/builders/238/builds/7702 broken, https://lab.llvm.org/buildbot/#/builders/238/builds/7703 green after cleanup
- [sanitizer-x86_64-linux](https://lab.llvm.org/buildbot/#/builders/37) - annotated builder: https://lab.llvm.org/buildbot/#/builders/37/builds/30519 broken, https://lab.llvm.org/buildbot/#/builders/37/builds/30521 green after two builds
- [sanitizer-aarch64-linux-bootstrap-asan](https://lab.llvm.org/buildbot/#/builders/239) - annotated builder, https://lab.llvm.org/buildbot/#/builders/239/builds/5498 broken due to merge-fdata target (related), https://lab.llvm.org/buildbot/#/builders/239/builds/5499 broken due to some other unrelated cause. Cleanup step resolved cmake issue.
- [premerge-monolithic-windows](https://lab.llvm.org/buildbot/#/builders/271): not annotated builder, https://lab.llvm.org/buildbot/#/builders/271/builds/3444 - broken as clean-build-dir step failed, https://lab.llvm.org/buildbot/#/builders/271/builds/3446 - green after successful cleanup.
- [libc-x86_64-debian-fullbuild-dbg](https://lab.llvm.org/buildbot/#/builders/163) - annotated builder: https://lab.llvm.org/buildbot/#/builders/163/builds/50669 - broken, still broken because of a **missing** cleanup step.
- [libc-x86_64-debian-dbg-runtimes-build](https://lab.llvm.org/buildbot/#/builders/225) - annotated builder, https://lab.llvm.org/buildbot/#/builders/225/builds/30309 broken, missing cleanup step.
- [clang-hip-vega20](https://lab.llvm.org/buildbot/#/builders/165) - annotated builder, https://lab.llvm.org/buildbot/#/builders/165/builds/48023 broken, missing cleanup step.
- [libc-x86_64-debian-gcc-fullbuild-dbg](https://lab.llvm.org/buildbot/#/builders/250) - annotated builder, https://lab.llvm.org/buildbot/#/builders/250/builds/17603 broken, missing cleanup step.
- [libc-aarch64-ubuntu-fullbuild-dbg](https://lab.llvm.org/buildbot/#/builders/223) - annotated builder, https://lab.llvm.org/buildbot/#/builders/223/builds/35258 broken, missing cleanup step.
- [sanitizer-windows](https://lab.llvm.org/buildbot/#/builders/127) - annotated builder, https://lab.llvm.org/buildbot/#/builders/127/builds/61496 broken, missing cleanup step.

In short, this diff is a red herring for the reported breakages. The real cause is annotated builders not handling build dir cleanup properly. Still, I'm reaching out to all of the broken buildbot owners asking to clean the build directory manually.

**Comment by aaupov** [comment]
This issue with incremental builds is explicitly mentioned in [HowToAddABuilder](https://llvm.org/docs/HowToAddABuilder.html):

> Use CCache and NOT incremental builds
Using ccache materially improves average build times. Incremental builds can be slightly faster, but introduce the risk of build corruption due to e.g. state changes, etc… At this point, the recommendation is not to use incremental builds and instead use ccache as the latter captures the majority of the benefit with less risk of false positives.

I've sent an email to builders owners. Still, sorry about landing it on a weekend. I expected that builders are configured according to the guidelines so the issues will clear automatically. Apologies for those depending on these broken builders.

**Comment by aaupov** [comment]
Status update:
- [sanitizer-windows](https://lab.llvm.org/buildbot/#/builders/127): back to green starting with https://lab.llvm.org/buildbot/#/builders/127/builds/61505
- [libc-aarch64-ubuntu-fullbuild-dbg](https://lab.llvm.org/buildbot/#/builders/223): green https://lab.llvm.org/buildbot/#/builders/223/builds/35300
- [libc-x86_64-debian-gcc-fullbuild-dbg](https://lab.llvm.org/buildbot/#/builders/250): green https://lab.llvm.org/buildbot/#/builders/250/builds/17643
- [libc-x86_64-debian-dbg-runtimes-build](https://lab.llvm.org/buildbot/#/builders/225): green https://lab.llvm.org/buildbot/#/builders/225/builds/30352
- [libc-x86_64-debian-fullbuild-dbg](https://lab.llvm.org/buildbot/#/builders/163): green https://lab.llvm.org/buildbot/#/builders/163/builds/50708
- [clang-hip-vega20](https://lab.llvm.org/buildbot/#/builders/165): red

**Comment by aaupov** [comment]
Final update:
- [clang-hip-vega20](https://lab.llvm.org/buildbot/#/builders/165): green https://lab.llvm.org/buildbot/#/builders/165/builds/48052
